packages/storage-plus/src/snapshot_map.rs

```diff
@@ -0,0 +1,386 @@
+#![cfg(feature = "iterator")]
+
+use serde::de::DeserializeOwned;
+use serde::Serialize;
+
+use cosmwasm_std::{Order, StdError, StdResult, Storage};
+
+use crate::keys::{EmptyPrefix, PrimaryKey, U64Key};
+use crate::map::Map;
+use crate::path::Path;
+use crate::prefix::Prefix;
+use crate::snapshot::{ChangeSet, Snapshot};
+use crate::{Bound, Prefixer, Strategy};
+
+/// Map that maintains a snapshots of one or more checkpoints.
+/// We can query historical data as well as current state.
+/// What data is snapshotted depends on the Strategy.
+pub struct SnapshotMap<'a, K, T> {
+    primary: Map<'a, K, T>,
+    snapshots: Snapshot<'a, K, T>,
+}
+
+impl<'a, K, T> SnapshotMap<'a, K, T> {
+    /// Usage: SnapshotMap::new(snapshot_names!("foobar"), Strategy::EveryBlock)
+    pub const fn new(
+        pk: &'a str,
+        checkpoints: &'a str,
+        changelog: &'a str,
+        strategy: Strategy,
+    ) -> Self {
+        SnapshotMap {
+            primary: Map::new(pk),
+            snapshots: Snapshot::new(checkpoints, changelog, strategy),
+        }
+    }
+}
+
+impl<'a, K, T> SnapshotMap<'a, K, T>
+where
+    T: Serialize + DeserializeOwned + Clone,
+    K: PrimaryKey<'a> + Prefixer<'a>,
+{
+    pub fn add_checkpoint(&self, store: &mut dyn Storage, height: u64) -> StdResult<()> {
+        self.snapshots.add_checkpoint(store, height)
+    }
+
+    pub fn remove_checkpoint(&self, store: &mut dyn Storage, height: u64) -> StdResult<()> {
+        self.snapshots.remove_checkpoint(store, height)
+    }
+}
+
+impl<'a, K, T> SnapshotMap<'a, K, T>
+where
+    T: Serialize + DeserializeOwned + Clone,
+    K: PrimaryKey<'a> + Prefixer<'a>,
+{
+    pub fn key(&self, k: K) -> Path<T> {
+        self.primary.key(k)
+    }
+
+    pub fn prefix(&self, p: K::Prefix) -> Prefix<T> {
+        self.primary.prefix(p)
+    }
+
+    pub fn sub_prefix(&self, p: K::SubPrefix) -> Prefix<T> {
+        self.primary.sub_prefix(p)
+    }
+
+    /// load old value and store changelog
+    fn write_change(&self, store: &mut dyn Storage, k: K, height: u64) -> StdResult<()> {
```

Comment:
I would refactor this and `may_load_at_height` to be in Snapshot, as I mentioned under Item

Reply:
Here again, the need for `primary` makes this unelegant to refactor. I would leave the current version.